!!! FEATURE: Publishing Version 3

mhsdesign opened the pull request:
Resolves: https://github.com/neos/neos-development-collection/issues/5327
Resolves: https://github.com/neos/neos-development-collection/issues/5303

Replaces: https://github.com/neos/neos-development-collection/pull/5300
Replaces: https://github.com/neos/neos-development-collection/pull/5302

**Introduces simulated publishing for rebasing commands and their constraint checks**

## Bugfixes

- emit $pointWorkspaceToNewContentStream event before applying remaining events in publish partial and rebase and partial discard
- do not emit events during simulating the rebase (e.g checking if there are conflicts)
- only publish all preserved the initiating timestamp, now a rebase or publish/discard individual nodes does the same
- proper exception strategy when the remainder of publish individual nodes cannot be applied (nothing gets published!)


## Publish workspace with automatic rebase for changes

We [discussed](https://neos-project.slack.com/archives/C04PYL8H3/p1729862662274719) whether a publish should only be allowed if the workspace is up to date, as technically a non conflicting change could sneak in like: The pages title was changed in live and the to be published text references the old title.

But this is only true in theory. It's more annoying having to rebase whenever multiple people work on potentially completely different parts of the site (or even a different site altogether).
To avoid the mentioned issue we could introduce at some point an indicator if someone else is working on the same document (or if that document has new changes in the target workspace). 

Allowing publish and publish individual nodes both even if the workspace was outdated optimises performance as both will now do a rebase as part of the publish strategy making an explicit rebase from the outside obsolete and thus speeding up the process a lot.

## Publish/Discard individual vs all #5303

Publishing all changes but using the individual publish (like the neos ui does) will now lead to the event `WorkspaceWasPublished` being emitted as this more performant to handle in the projections.

Similarly we do the same with discard individual and discard all

## Publish/Discard/Rebase as no-op if there are no changes

- the individual publish / discard will directly return if there are no nodes specified
- if there are no changes publish and publish individual is a no-op and will just reopen the content stream
- if there are no things to be discarded discard (discard individual) is a no-op

## Performance old vs new

We also [discussed](https://neos-project.slack.com/archives/C04PYL8H3/p1729868353001999) if the performance of leveraging the db's rollback is acceptable:

| Count of Nodes | Time to create | Time for simulated publish | Time for previous publish    | Time for previous rebase & publish |
|----------------|----------------|----------------------------|------------------------------|---------------------------------------|
| 100            | 0.5s           | 0.4s                       |                              |                                       |
| 1000           | 4.6s           | 3.7s                       |                              |                                       |
| 5000           | 24s            | 23s                        |                              |                                       |
| 10000          | 46s - 49s      | 83s                        | 24s (unlikely if up-to-date) | 192s (rebase) + 176s (publish all) = 368s |

Tested with 1G memory limit but behat seems to have only used around 150Mb each run no matter how many nodes.

It seems that the publish is even slightly faster than the creation of the nodes until we exceed 5.000 newly created nodes. With 10.000 nodes it takes twice as much time.

The previous publish implementation wich can only work without changes in the target only copied the events which was quite performant, but as this is the edge-case as one needs to be up to date AND publish all changes (across sites) this optimisation was removed in favour of a more consistent behaviour that also is able to rebase if required.

The Neos Ui doesnt use the publish all but rather an explicit rebase coupled with a publish individual nodes. This is super slow and we are able to optimise this a lot by doing the rebase and publish in one step.

For the main use case the new implementation is not worse but even better than the old one as the explict rebase is extremely slow in combination with the the publish afterwards.

Disclaimer: The measurements are to be taken with a grain of salt. Christian also mentioned that the performance possibly gets worse at one point due to all the events being stored in memory (his 100k publication crashed due to memory limitation)


## Implementation of the `CommandSimulator`

The CommandSimulator is used during the publishing process, for partial publishing and workspace rebasing.

For this case, we want to apply commands including their constraint checks step by step, to see whether this
set of commands applies cleanly without errors, and which events would be created by them, but we do NOT
want to commit the updated projections or events.

Internally, we do the following:
- Create a database transaction in the GraphProjection which we will roll back lateron (to dry-run
projection updates) (via `CommandSimulator::run()`).
- Create an InMemoryEventStore which buffers created events by command handlers.
- execute all commands via `CommandSimulator::handle()`
- -> this will do all constraint checks based on the projection in the open transaction (so it sees
previously modified projection state which is not committed)
- -> it will run the command handlers, buffer all emitted events in the InMemoryEventStore
-> note to avoid full recursion the workspace command handler is not included in the bus
- -> update the GraphProjection, but WITHOUT committing the transaction `ContentGraphProjectionInterface::inSimulation()`

This is quite performant because we do not need to fork a new content stream

## Implementation of graph projection `inSimulation($fn)`

We introduce a new dedicated method `inSimulation` for simulated rebasing on the `ContentGraphProjectionInterface` (the only projection needed for soft constraint checks)

The implementation must ensure that the function passed is invoked and that any changes via `ContentGraphProjectionInterface::apply()` are executed "in simulation" e.g. NOT persisted after returning.

The projection state `ContentGraphReadModelInterface` must reflect the current changes of the simulation as well during the execution of the function.

For the doctrine content graph projection this is done by leveraging a transaction and rollback.

We [discussed](https://github.com/neos/neos-development-collection/pull/5301#issuecomment-2433457506) if relying on the global `Connection` (wired via flows and the entity manger) is safe as it must be ensured that NO one can hook into the process during rebase and commit the simulated state, especially not our own code in the projection or sub classes. By leveraging `$connection->setRollbackOnly()` (which will be unset after the rollback) we can exactly do that which is of course only a slim layer but gives enough insurance and doesn't come with the costs of having too many dedicated connections.



**Upgrade instructions**

Migration will be possible via https://github.com/neos/neos-development-collection/pull/5297

```
./flow migrateevents:backup
./flow contentStream:removeDangling
./flow contentStream:pruneRemovedFromEventStream
./flow cr:projectionReplayAll
```

**Review instructions**

_Relation to other changes:_

1.) Requires https://github.com/neos/neos-development-collection/pull/5272 as the content graph projection is required as special projection for the simulation with a `inSimulation` method.

2.) Based on https://github.com/neos/neos-development-collection/pull/5315 to introduce `yield` for command handlers

3.) As part of the migration we require the content streams to be pruned. This is currently unstable and will be fixed with the content stream pruner overhaul followup: https://github.com/neos/neos-development-collection/pull/5297. The change will also remove the left-over content stream commands that are not required by the workspace command handler anymore.


<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

kitsunet committed:
BUGFIX: Partial publish breaks uri and change Projection

A partial publish results in events annotated for "live"
workspace yet are not in the "live" content stream.

This is due to a fork of the live content stream being
created with the partially published events in, which is
then published to the actual live content stream.
This leaves behind duplicate events both containing
"workspaceName: live" yet only one of them is in the live
content stream. A catchup or replay will fail however due
to duplicate database entries for the reflections relying
on anything with "workspaceName: live" being actually in
the live content stream.

The provided test fails showing the behavior.

mhsdesign committed:
WIP task/schnappsidee-rebase-and-partial-publish-in-memory-event-stream

mhsdesign committed:
WIP: Commit for command handling in inSimulation

mhsdesign committed:
WIP: handlePublishIndividualNodesFromWorkspace with remaining events also being temporary

mhsdesign committed:
WIP: move handleRebaseWorkspace to run inSimulation

mhsdesign committed:
WIP: refactor handleDiscardIndividualNodesFromWorkspace to run inSimulation

mhsdesign committed:
WIP todos

kitsunet commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -38,13 +48,93 @@ final class CommandHandlingDependencies
+        }
+    }
+
+    private function hanldeInSimulation(CommandInterface $command, InMemoryEventStore $inMemoryEventStore): CommandResult
```
handle

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/schnappsidee-rebase-and-partial-publish-in-memory-event-stream

mhsdesign committed:
TASK: Adjustments after merge of cr first level projection

mhsdesign committed:
Merge branch 'task/schnappsidee-zwo-yield-events-to-publish-in-workspace-command-handler' into task/schnappsidee-rebase-and-partial-publish-in-memory-event-stream

mhsdesign committed:
TASK: Introduce CommandSimulator and refactor publish partially and rebase to use yield

Also the object wiring was cleaned up and simplified.
A full recursion is now no longer but was never needed: Eg rebase a CreateWorkspace command or the likes.

skurfuerst commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/W6-WorkspaceRebasing/02-RebasingWithAutoCreatedNodes.feature:
```diff
@@ -69,8 +69,13 @@ Feature: Rebasing auto-created nodes works
+    # user ws must be outdated!!! otherwise cheesy
+
+    # rebase of SetSerializedNodeProperties
+    # More than one node anchor point for content stream: b2a1d336-38a7-4183-815c-c49b2c7eba8c, node aggregate id: 7ef3c166-16d5-4aed-b8cb-5a44670607b7 and origin dimension space point: [] – this should not happen and might be a conceptual problem!
```
need to discuss

mhsdesign replied:
we removed the todos and made sure the test is never a noop

skurfuerst commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/W8-IndividualNodePublication/03-MoreBasicFeatures.feature:
```diff
@@ -134,40 +134,40 @@ Feature: Publishing individual nodes (basics)
-    And I expect this node to have the following properties:
-      | Key   | Value            |
-      | image | "Modified image" |
+#  Scenario: It is possible to publish no node
```
wieder einkommentieren, oder weghauen?

mhsdesign replied:
the assertions should change imo, now its a noop thats why we comment it out first ...

mhsdesign committed:
TASK: Remove now obsolete `ForkContentStream` command

mhsdesign committed:
TASK: Adjust todo comments

skurfuerst commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,125 @@
+use Neos\EventStore\Model\EventStream\VirtualStreamName;
+
+/**
+ * Implementation detail of {@see ContentRepository::handle}, when rebasing or partially publishing
```
```
The CommandSimulator is used during the publishing process, for partial publishing and workspace rebasing.

For this case, we want to apply commands including their constraint checks step by step, to see whether this
set of commands applies cleanly without errors, and which events would be created by them, but we do NOT
want to commit the updated projections or events.

Internally, we do the following:
- Create a database transaction in the GraphProjection which we will rollback lateron (to dry-run
  projection updates) (via {@see run()}).
- Create an InMemoryEventStore which buffers created events by command handlers.
- execute all commands via {@see handle()}
- -> this will do all constraint checks based on the projection in the open transaction (so it sees
  previously modified projection state which is not committed)
- -> it will run the command handler, buffer all emitted events in the InMemoryEventStore
- -> update the GraphProjection, but WITHOUT committing the transaction.

This is quite performant because we do not need to fork a new content stream.
```

kitsunet replied:
still makes sense as additional context comment

mhsdesign replied:
yeah added thank you ❤️ ;)

skurfuerst commented on Neos.ContentRepository.Core/Classes/CommandHandler/ControlFlowAwareCommandHandlerInterface.php:
```diff
@@ -0,0 +1,27 @@
+use Neos\ContentRepository\Core\EventStore\EventsToPublishFailed;
+
+/**
+ * Common interface for all Content Repository command handlers
```
interface for all Content Repository command handlers which need to interact with the control flow
of event publishing, e.g. for error handling.
This is used in combination with CommandSimulator for Workspace partial Publish and rebase.

kitsunet replied:
still makes sense as additional context comment

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -259,6 +259,19 @@ public function apply(EventInterface $event, EventEnvelope $eventEnvelope): void
 
+    public function inSimulation(\Closure $fn): mixed
+    {
+        if ($this->dbal->isTransactionActive()) {
```
did you try to get the `setRollbackOnly()` working (potentially on a clone of the $dbal instance)?
Otherwise I would fear that some code somewhere commits the transaction

mhsdesign replied:
i mean we _could_ set it to true and then use reflection to unset it again 😅 
but i dont see a way how we can use a clone here ... except if we were to set the clone on $this->dbal = $clone ... but that would also be ugly?

mhsdesign replied:
maybe we can wrap the instance in a decorator on creation of the whole projection?

kitsunet replied:
we decided in a discussion to leave it be for the moment and come back to how we want to split / protect the connectoin at a later point.

mhsdesign committed:
TASK: Adjust test 02-RebasingWithAutoCreatedNodes.feature

by creating live changes we ensure that the rebase is never a noop

mhsdesign committed:
TASK: Adjust exceptions to be a noop if PublishIndividualNodesFromWorkspace doesnt need to publish

mhsdesign committed:
TASK: Introduce test for quick path if rebased commands emitted 0 events

... and that the user content stream did not change and is still open!

mhsdesign committed:
TASK: Add quick paths for partial discarding with tests

mhsdesign committed:
Add todos to optimise partial publish to full publish and discard all

mhsdesign committed:
TASK: Make publish all a no-op if there are no changes and also ensure the content stream is closed

mhsdesign committed:
TASK: Remove PublishIndividualNodesFromWorkspace::contentStreamIdForMatchingPart

... as we dont do two forks anymore but try to handle the commands in simulation

mhsdesign committed:
TASK: Remove outdated comments and adjust variable namings

kitsunet committed:
Workspace aware simulation

kitsunet commented:
Do we see any risks in that the _one_ connection used here is shared between ORM and eventstore and the CR? Given we _must_ ensure nothing commits the transaction we started for the simulation. So no hooks, signals or anything that might lead to userland code being executed that could in some way start/commit a transaction.

mhsdesign committed:
Cosmetics & Nagellack

bwaidelich commented on Neos.ContentRepository.BehavioralTests/Classes/Command/PerformanceMeasurementService.php:
```diff
@@ -154,9 +153,10 @@ private function createHierarchy(
+        // $this->contentRepository->handle(ForkContentStream::create(
+        //     ContentStreamId::create(),
+        //     $this->contentStreamId,
+        // ));
```
as discussed with @kitsunet we should probably remove the whole service for now (in a separate PR so that we can revert once the lowlevel changes are done)
Christian is working on a different approach for benchmarks and will steal some ideas from this class

bwaidelich commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/01-ForkContentStream_ConstraintChecks.feature:
```diff
@@ -49,12 +49,13 @@ Feature: ForkContentStream Without Dimensions
       | propertiesToUnset            | {}                                                      |
 
-  Scenario: Try to fork a content stream that is closed:
+  Scenario: Try to fork a content stream (use it as base workspace) that is closed:
```
```suggestion
  Scenario: Try to create a workspace based on a workspace that has a closed content stream
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        $this->inSimulation = true;
+        try {
+            return $this->contentRepositoryProjection->inSimulation($fn);
```
Suggestion: Make `handle` private and change this to:
```suggestion
            return $this->contentRepositoryProjection->inSimulation($fn($this->handle(...));
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,140 @@
+    {
+        if ($this->inSimulation === false) {
+            throw new \RuntimeException('Simulation is not running');
+        }
```
With the suggestion above we could skip this check

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,140 @@
+            throw new \RuntimeException('Simulation is not running');
+        }
+
+        // FIXME: Check if workspace already matches and skip this
```
```suggestion
        // FIXME: Check if workspace already matches and skip this, e.g. $commandInWorkspace = $command->getWorkspaceName()->equals($this->workspaceNameToSimulateIn) ? $command : $command->createCopyForWorkspace($this->workspaceNameToSimulateIn);
        // when https://github.com/neos/neos-development-collection/pull/5298 is merged
```

kitsunet replied:
still makes sense as additional context comment

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,140 @@
+
+    public function currentSequenceNumber(): SequenceNumber
+    {
+        foreach ($this->inMemoryEventStore->load(VirtualStreamName::all())->backwards()->limit(1) as $eventEnvelope) {
```
```suggestion
        foreach ($this->eventStream()->backwards()->limit(1) as $eventEnvelope) {
```

kitsunet replied:
still makes sense

mhsdesign committed:
TASK: Remove obsolete check to skip events

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -101,38 +100,64 @@ public function handle(CommandInterface $command): void
-        $this->eventPersister->publishEvents($this, $eventsToPublish);
+        $eventsToPublishOrGenerator = $this->commandBus->handle($command, $this->commandHandlingDependencies);
+
+        // TODO only apply metadata if is copied event?? In generator below we ignore this...
```
my bad: we already only set metadata if they didn't exist so we can safely do this in both cases.

We should extract the common code to a method

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php:
```diff
@@ -17,16 +17,19 @@
 {
     /**
-     * @var CommandHandlerInterface[]
+     * @var (CommandHandlerInterface|ControlFlowAwareCommandHandlerInterface)[]
```
IMO we don't need the additional interface because we already check for the return type in `ContentRepository::handle()`

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,140 @@
+    private function handleCommand(CommandInterface $command, CommandHandlingDependencies $commandHandlingDependencies): EventsToPublish
+    {
+        // TODO fail if multiple handlers can handle the same command
+        foreach ($this->handlers as $handler) {
```
use `CommandBus`

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ProjectionInterface.php:
```diff
@@ -47,4 +47,6 @@ public function getCheckpointStorage(): CheckpointStorageInterface;
     public function getState(): ProjectionStateInterface;
 
     public function reset(): void;
+
+    // public function temporaryChanges(\Closure $fn): void;
```
```suggestion
```

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/schnappsidee-rebase-and-partial-publish-in-memory-event-stream

mhsdesign committed:
TASK: Remove extra interface for workspace command handler and combine it with the default handler interface

also use the command bus directly and just throw in the simulation when the - never going to happen case - of returning a generator occurs.

mhsdesign committed:
TASK: Use `currentSequenceNumber` instead of calculating it ourselves

to make it more readable

mhsdesign committed:
BUGFIX: Move workspace pointer BEFORE remaining events are emitted

Previously the asset usage catchup hook would fail

> Type error: Neos\Neos\AssetUsage\Service\AssetUsageIndexingService_Original::removeIndexForNode(): Argument #2 ($node) must be of type Neos\ContentRepository\Core\Projection\ContentGraph\Node, null given

because the workspace name would still point to the previous one while the events already happen on the new stream. Now we fully ensure that we repoint the workspace first before applying the remainder.

And fix of-by-one error by no fetching too many remaining events see `$highestVersionForMatching->previous()`

mhsdesign committed:
TASK: Remove old content-stream after rebase

mhsdesign committed:
TASK: Ensure that dangling content streams are removed after publish discard and rebase

mhsdesign committed:
TASK: Simplify CommandSimulator by exposing handle function only during run

mhsdesign committed:
TASK: Adjust content stream prune test as rebase is a noop without changes

the force strategy ensures to always fork a new content stream

mhsdesign committed:
TASK: Always rebase during publication and do not copy events directly

kitsunet committed:
DimensionSpacePoint events enriched

kitsunet commented:
Alright, at least according to the tests this works now. I also went through all Rebasable commands to check if the events get enriched. the Dimension ones were the only missing. IMHO we _should_ centralize this behavior, I opted against doing it here though as I think we need to consider if there would have to be any more logic involved to decide what gets enriched with commands, in what way, when we override the command if there is already metadata and finally what the causation ids are. I _guess_ we could ignore all these questions and centralize it, but it warrants a closer look and is therefore out of scope of this change.

kitsunet committed:
Just some linter happiness

kitsunet reviewed (approved):
Loose +1 from me, I think this is in pretty good shape. But as I had so much influence on the change, my +1 should be taken with a grain of salt.

mhsdesign committed:
TASK: Consistently use workspace instead of content graph in workspace command handler

... and remove obsolete todos (as tests do fail now)

mhsdesign committed:
TASK: Introduce `InitiatingEventMetadata` utility to centralise access and use

mhsdesign committed:
TASK: Introduce ExtractedCommand vo to collect original commands AND their initiating metadata

mhsdesign committed:
TASK: Combine `NodeAggregateEventPublisher` and `ExtractedCommand` to `RebasableCommand`

... because the name `NodeAggregateEventPublisher` is incorrect as we need to enrich also dimension changed events and to solve the todo:

> TODO: Add this logic to the NodeAggregateCommandHandler;
> so that we can be sure these can be parsed again.

mhsdesign committed:
TASK: Add test that rebase keeps the originalCreated intact

see $rebaseableCommand->initiatingMetaData in 29a27deaf72caac9ddc319764efffb8c3a4a386b

mhsdesign committed:
TASK: Move `CommandHandlingDependencies` to CommandHandler namespace

mhsdesign committed:
TASK: Dont pass `CommandHandlingDependencies` through half of the world

... at least not through the CR

mhsdesign committed:
TASK: Update todos and require base content stream to be not closed during publishing

... doing a full publish in $remainingCommands would be more complicated and expensive as we would need to refetch the evens

mhsdesign committed:
TASK: Remove obsolete `ContentGraphReadModelInterface::buildContentGraph`

We intentionally got rid of all usages to create a "custom" content graph with outside defined content stream id

mhsdesign committed:
TASK: Turn partial publish into full publish if all nodes are about to be published

... this should speed up the projections especially as the Neo ui always only does a partial publish

Also publish will be a rebase if the workspace is outdated and there are no changes.

mhsdesign committed:
TASK: Update documentation

mhsdesign committed:
TASK: Revert currently unused error handling strategy from workspace command handler

Before 29a27deaf72caac9ddc319764efffb8c3a4a386b we required the ->throw() behaviour of https://github.com/neos/neos-development-collection/pull/5315 to catch the `ConcurrencyException` inside in the command handler.

mhsdesign committed:
FEATURE: Throw `WorkspaceRebaseFailed` during publication or partial discard as well

... instead of enforcing to rebase from outside.
This speeds things up as we need one fork less;)

see also https://github.com/neos/neos-development-collection/pull/5300

mhsdesign committed:
BUGFIX: Dont discard remaining user changes if $matchingCommands is empty

... and revert doing just a reopen when the matching part doesnt yield any events (as no rebase would be excecuted then)

mhsdesign commented on Neos.ContentRepository.Core/Classes/EventStore/InitiatingEventMetadata.php:
```diff
@@ -0,0 +1,69 @@
+/**
+ * @internal
+ */
+final readonly class InitiatingEventMetadata
```
new utility to encapsulate access instead of passing magic strings around :)

see https://github.com/neos/neos-development-collection/pull/5301/commits/4ee595ebde08bce4686fd185b4d712c5c0555e76

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/RebaseableCommand.php:
```diff
@@ -2,34 +2,61 @@
  * @internal
  */
-final class NodeAggregateEventPublisher
+final readonly class RebaseableCommand
```
utility to pass the "extracted" commands from the event stream around and also enrich them...

... because the use of `NodeAggregateEventPublisher` is incorrect now as we need to enrich also dimension changed events ...

Also by extraction and enrichment into one place we solve this todo :)

> TODO: Add this logic to the NodeAggregateCommandHandler;
> so that we can be sure these can be parsed again.


see https://github.com/neos/neos-development-collection/pull/5301/commits/56fd230780dd0c23c3308ab1f8a0b250c1aa29f0 and https://github.com/neos/neos-development-collection/pull/5301/commits/f7c39ec64e0392b51b89321369dc204e71f6a80e

kitsunet replied:
Very nice

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceRebase/CommandThatFailedDuringRebase.php:
```diff
@@ -15,19 +15,20 @@
      */
     public function __construct(
-        public int $sequenceNumber,
+        public SequenceNumber $sequenceNumber,
```
todo this is api and will be interpreted in the neos ui, do we want to pass the `Neos\EventStore\Model\Event\SequenceNumber` to the outside?

see use: https://github.com/neos/neos-ui/blob/7468ffb851ac16713f8afda9277204a396f09a1c/Classes/Infrastructure/ContentRepository/ConflictsFactory.php#L118

kitsunet replied:
The CR is already "only" a user of the SequenceNumber, I find it a bit smelly to hand this around, but then it's not exactly internal to the CR either.

mhsdesign replied:
made it `@internal` and removed use in Neos ui ;) https://github.com/neos/neos-ui/pull/3769/commits/15c13ab601a44b54f178304558e48a58cb7ef1b8

mhsdesign commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,74 @@
+ *
+ * @internal
+ */
+final readonly class CommandHandlingDependencies
```
moved to `CommandHandler` namespace :)

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/W8-IndividualNodePublication/03-MoreBasicFeatures.feature:
```diff
@@ -155,27 +158,73 @@ Feature: Publishing individual nodes (basics)
+    And I expect this node to be exactly explicitly tagged "tag1"
+
+    When I am in workspace "user-test" and dimension space point {}
+    Then I expect node aggregate identifier "sir-unchanged" to lead to node user-cs-identifier-remaining;sir-unchanged;{}
```
@kitsunet i removed our special handling again when the events that were generated are empty as this would make the code way to complex for no gain except that theoretical case ... and here due to the live change the workspace was outdated and needs a rebase either way ... so impossible ^^
The expected content stream is now `user-cs-identifier-remaining` and not unchanged!

see https://github.com/neos/neos-development-collection/pull/5301/commits/49d58709e9a40c90ff7abb69132f3cf972f4a103

kitsunet replied:
Fine by me :)

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/Workspaces/PruneContentStreams.feature:
```diff
@@ -41,6 +41,7 @@ Feature: If content streams are not in use anymore by the workspace, they can be
     When the command RebaseWorkspace is executed with payload:
       | Key           | Value       |
       | workspaceName | "user-test" |
+      | rebaseErrorHandlingStrategy | "force"               |
```
for these tests ... as we dont make changes live and rebase is now a noop if the workspace is not outdated, we made "force" reflect the old behaviour to always fork even if there are no changes in the base.

mhsdesign commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,167 @@
+        // when https://github.com/neos/neos-development-collection/pull/5298 is merged
+        $commandInWorkspace = $rebaseableCommand->originalCommand->createCopyForWorkspace($this->workspaceNameToSimulateIn);
+
+        try {
+            $eventsToPublish = $this->commandBus->handle($commandInWorkspace);
+        } catch (\Exception $exception) {
+            $this->commandsThatFailedDuringRebase = $this->commandsThatFailedDuringRebase->withAppended(
+                new CommandThatFailedDuringRebase(
+                    $rebaseableCommand->originalSequenceNumber,
+                    $rebaseableCommand->originalCommand,
+                    $exception
+                )
+            );
+
+            return;
+        }
```
as this got really repetitive over the 5 usages i added the logic to collect the failures to the command simulator.

```php
$commandSimulator = $this->commandSimulatorFactory->createSimulator($baseWorkspace->workspaceName);

$commandSimulator->run(
    static function ($handle) use ($rebaseableCommands): void {
        foreach ($rebaseableCommands as $rebaseableCommand) {
            $handle($rebaseableCommand);
        }
    }
);

if ($commandSimulator->hasCommandsThatFailed()) {
    throw WorkspaceRebaseFailed::duringPublish($commandSimulator->getCommandsThatFailed());
}
```

i thought of returning it first but that would mean that the return value of `$handle` has to be manually collected or we create a lot of logic in the `run` closure ... as we do the same with `$commandSimulator->eventStream()` i think its fine to do:) The object has state already ^^

kitsunet replied:
Does it have state? I thought by putting handle as closure we got rid of all the state? But anyways looks good.

mhsdesign replied:
not _itself_ ... it was `readonly` but as the class contains and exposes the in-memory event store this is really stateful already ;)

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspacePublishingService.php:
```diff
@@ -255,19 +276,15 @@ private function discardNodes(
         ContentRepository $contentRepository,
         WorkspaceName $workspaceName,
         NodeIdsToPublishOrDiscard $nodeIdsToPublish
     ): void {
-        /**
-         * TODO: only rebase if necessary!
-         * Also, isn't this already included in @see WorkspaceCommandHandler::handlePublishIndividualNodesFromWorkspace ?
-         */
-        $contentRepository->handle(
-            RebaseWorkspace::create($workspaceName)
-        );
-
         $contentRepository->handle(
             PublishIndividualNodesFromWorkspace::create(
```
We throw now the `WorkspaceRebaseFailed` in case one of the commands could not be handled, previously we would just throw the raw exception in `PublishIndividualNodesFromWorkspace` instead.

The `WorkspaceRebaseFailed` was only due to the now obsolete `RebaseWorkspace` here which is required for the ui to react accordingly: https://github.com/neos/neos-ui/pull/3769

see https://github.com/neos/neos-development-collection/pull/5301/commits/8ad0a3e5c5f968f8b8b1fda83b3df6710834b95b

mhsdesign reviewed (commented):
i commented the bits and pieces where you guys might want to add your last bit of mustard to:) (the changes of the last two days without christians help xd)

kitsunet commented:
I am still happy with this, thanks for doing the extra mile.

mhsdesign committed:
TASK: Minor cosmetic adjustments and documentation

mhsdesign committed:
TASK: Use `setRollbackOnly` to ensure nothing is commited during the simulation

solves https://github.com/neos/neos-development-collection/pull/5301#issuecomment-2433457506

mhsdesign committed:
TASK: Make publish a no-op if there are no changes instead of a rebase

see discussion https://neos-project.slack.com/archives/C04PYL8H3/p1730111291595599

otherwise the logic got way to complex :D

previously a publishing without changes is a no-op if up to date and rebase if outdated

mhsdesign committed:
TASK: Turn full discard also into a no-op if there are no changes

mhsdesign committed:
TASK: Adjust comments to not reference renamed `NodeAggregateEventPublisher`

mhsdesign committed:
TASK: Make `CommandsThatFailed::sequenceNumber` internal only for testing

see https://github.com/neos/neos-development-collection/pull/5301#discussion_r1818111183

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php:
```diff
@@ -4,38 +4,38 @@
      * @var CommandHandlerInterface[]
      */
-    private array $handlers;
+    public array $handlers;
```
Not a blocker but I would prefer the handlers not to be publicly exposed here.
AFAIS this is only used in the CR factory:
```php
$publicCommandBus = new CommandBus(
    $commandHandlingDependencies,
    new ContentStreamCommandHandler(),
    new WorkspaceCommandHandler(
        $commandSimulatorFactory,
        $this->projectionFactoryDependencies->eventStore,
        $this->projectionFactoryDependencies->eventNormalizer,
    ),
    ...$commandBusForRebaseAbleCommands->handlers
);
```
but maybe we can do something like
```php
$publicCommandBus = $commandBusForRebaseAbleCommands->withAdditionalHandlers(
  $commandHandlingDependencies,
  new ContentStreamCommandHandler(),
  new WorkspaceCommandHandler(
      $commandSimulatorFactory,
      $this->projectionFactoryDependencies->eventStore,
      $this->projectionFactoryDependencies->eventNormalizer,
  ),
);
```

instead to keep it private?
(I know the CommandBus is `@internal` but...?

mhsdesign replied:
hmm i agree... we only need it for the factory ... 3rd option store them in a array and use that ... ill do that:)

mhsdesign replied:
nevermind `withAdditionalHandlers ` works better :D

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php:
```diff
@@ -4,38 +4,38 @@
+    public function handle(CommandInterface $command): EventsToPublish|\Generator
     {
-        // TODO fail if multiple handlers can handle the same command
+        // multiple handlers must not handle the same command
```
You just renamed the phrase and removed the TODO, sneaky :)
I think it is fine like this, but we could also consider sth along the lines of:
```php
$resolvedHandlers = array_filter($this->handlers, fn (CommandHandlerInterface $handler) => $handler->canHandle($command));
if ($resolvedHandlers === []) {
  // throw no handler found
}
if (count($resolvedHandlers) > 1) {
  // throw multiple handlers found
}
return $resolvedHandlers[0]->handle($command, $this->commandHandlingDependencies);
```

mhsdesign replied:
i had it fixed at some point while moving stuff around and well the exception is rightfully never thrown ... and then i thought why have the overhead of calling `canHandle` that many times if we can just stop at the first match (especially they use reflection internally and stuff) so i thought that for us core developers it wouldnt be much of a gain as there are a lot more places where stuff _could_ fail when creating new commands ... so i think its fine to not guard it here

bwaidelich replied:
Fine with me!

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlerInterface.php:
```diff
@@ -20,6 +18,11 @@ interface CommandHandlerInterface
     /**
+     * "simple" command handlers return EventsToPublish directly
+     *
+     * For the case of the workspace command handler who need to publish to many streams and "close" the content-stream directly,
```
```suggestion
     * For the case of the workspace command handler that need to publish to many streams and "close" the content-stream directly,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,74 @@
+
+    public function contentStreamExists(ContentStreamId $contentStreamId): bool
+    {
+        $cs = $this->contentGraphReadModel->findContentStreamById($contentStreamId);
+        return $cs !== null && !$cs->removed;
```
nitpick but according to our CGL:
```suggestion
        $contentStream = $this->contentGraphReadModel->findContentStreamById($contentStreamId);
        return $contentStream !== null && !$contentStream->removed;
```

And related to that: Wasn't the plan to get rid of the `removed` flag? (or is that part of a separate PR?)

mhsdesign replied:
yeah this line will be fully eradicated via https://github.com/neos/neos-development-collection/pull/5297 so im not applying the cgl thing :D

bwaidelich replied:
Sweet!

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandler/CommandSimulator.php:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        if (!$eventsToPublish instanceof EventsToPublish) {
+            throw new \RuntimeException(sprintf('CommandSimulator expects direct EventsToPublish to be returned when handling %s', $rebaseableCommand->originalCommand::class));
```
nitpick
```suggestion
            throw new \RuntimeException(sprintf('%s expects an instance of %s to be returned. Got %s when handling %s', self::class, EventsToPublish::class, get_debug_type($eventsToPublish), $rebaseableCommand->originalCommand::class));
```

bwaidelich reviewed (approved):
I admire the amount of work you have put into this, thank you!

I don't feel qualified to do a proper review of the inner workings. But the parts that I understood make a lot of sense to me and the complexity is not that much higher – and you added quite a lot of tests, too.

So I'd say, let's get this merged asap so that we can test it further.

I just added some rather nitpicky comments, feel free to ignore those

mhsdesign committed:
Apply suggestions from code review

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

mhsdesign committed:
TASK: Declare `$handlers` of `CommandBus` private
